core/internal/runmetadata/runmetadata.go

```diff
@@ -351,6 +377,66 @@ func (metadata *RunMetadata) signalDirty() {
 	}
 }
 
+// updateMetadataForResume updates run metadata based on the existing run
+// that's being resumed.
+func (metadata *RunMetadata) updateMetadataForResume(
+	resumeSetting string,
+) error {
+	if metadata.graphqlClientOrNil == nil {
+		// Ignore the resume mode when offline.
+		//
+		// A warning is printed by the client during wandb.init().
+		//
+		// resume="auto" is always OK and is handled by the client.
+		return nil
+	}
+
+	return runbranch.NewResumeBranch(
+		metadata.beforeRunEndCtx,
+		metadata.graphqlClientOrNil,
+		resumeSetting,
+	).UpdateForResume(
+		metadata.params,
+		metadata.config,
+	)
+}
+
+// updateMetadataForRewind updates run metadata based on the existing run
+// that's being rewound.
+func (metadata *RunMetadata) updateMetadataForRewind(
+	rewindSetting *spb.BranchPoint,
+) error {
+	if metadata.graphqlClientOrNil == nil {
+		return &RunUpdateError{
+			UserMessage: "Cannot use `resume_from` in an offline run.",
+			Cause:       errors.New("runmetadata: cannot rewind a run when offline"),
+			Code:        spb.ErrorInfo_USAGE,
```

Comment:
"Cannot use `resume_from` in an offline run." Why not? People actually do exactly that, when they are in an isolated environment (see https://github.com/wandb/wandb/pull/9876), or this just a sanity check?
it might error out when they try to sync, of course.

Reply:
If we're offline, we don't know the starting step, so all the history entries in the transaction log are wrong. You can try it out, it doesn't work correctly:

```python
with wandb.init() as run:
    run.log({"x": 1})
    run.log({"x": 2})
    run.log({"x": 3})


with wandb.init(resume_from=f"{run.id}?_step=1", mode="offline") as run:
    run.log({"x": 4})
```

You can check its transaction log using `wandb sync --view --verbose`. Note that the history entry's "_step" is 0.

After you sync, the summary shows `x: 4`, but the workspace view shows the points `{"_step": 0, "x": 1}` (correct) and `{"_step": 0, "x": 2}` (wrong, should have `"x": 4`).

Reply:
you're right, in offline mode we should just treat the requested step value as if it were correct and start from it. the backend will do the actual correctness checks on sync (and throw an error if something is wrong).